tests/Transports.AspNetCore.Tests/ChatTests.cs

```diff
@@ -0,0 +1,445 @@
+        await AddMessageInternal(2);
+
+        var str = await _app.ExecuteGet("/graphql?query={allMessages{id message}}");
+        str.ShouldBe("{\"data\":{\"allMessages\":[{\"id\":\"1\",\"message\":\"hello\"},{\"id\":\"2\",\"message\":\"hello\"}]}}");
+    }
+
+    public async Task AddMessageInternal(int id, string name = "John Doe")
+    {
+        var str = await _app.ExecutePost(
+            "/graphql",
+            "mutation {addMessage(message:{message:\"hello\",from:\"" + name + "\"}){id}}");
+        str.ShouldBe("{\"data\":{\"addMessage\":{\"id\":\"" + id + "\"}}}");
+    }
+
+    [Fact]
+    public async Task DeleteMessage()
+    {
+        await AddMessageInternal(1);
+        await AddMessageInternal(2);
+
+        var str = await _app.ExecuteGet("/graphql?query={count}");
+        str.ShouldBe("{\"data\":{\"count\":2}}");
+
+        await DeleteMessageInternal(1);
+
+        str = await _app.ExecuteGet("/graphql?query={count}");
+        str.ShouldBe("{\"data\":{\"count\":1}}");
+    }
+
+    public async Task DeleteMessageInternal(int id)
+    {
+        var str = await _app.ExecutePost(
+            "/graphql",
+            "mutation {deleteMessage(id:\"" + id + "\"){id}}");
+        str.ShouldBe("{\"data\":{\"deleteMessage\":{\"id\":\"" + id + "\"}}}");
+    }
+
+    [Fact]
+    public async Task ClearMessages()
+    {
+        await AddMessage();
+        await ClearMessagesInternal(1);
+    }
+
+    private async Task ClearMessagesInternal(int numDeleted)
+    {
+        var str = await _app.ExecutePost(
+            "/graphql",
+            "mutation {clearMessages}");
+        str.ShouldBe("{\"data\":{\"clearMessages\":" + numDeleted + "}}");
+    }
+
+    [Theory]
+    [InlineData("graphql-ws")]
+    [InlineData("graphql-transport-ws")]
+    public async Task Subscription(string subProtocol)
+    {
+        // create websocket connection
+        var webSocketClient = _app.CreateWebSocketClient();
+        webSocketClient.ConfigureRequest = request =>
+        {
+            request.Headers["Sec-WebSocket-Protocol"] = subProtocol;
+        };
+        webSocketClient.SubProtocols.Add(subProtocol);
+        using var webSocket = await webSocketClient.ConnectAsync(new Uri(_app.BaseAddress, "/graphql"), default);
+
+        // send CONNECTION_INIT
+        await webSocket.SendMessageAsync(new OperationMessage
+        {
+            Type = "connection_init"
+        });
+
+        // wait for CONNECTION_ACK
+        var message = await webSocket.ReceiveMessageAsync();
+        message.Type.ShouldBe("connection_ack");
+
+        // subscribe
+        await webSocket.SendMessageAsync(new OperationMessage
+        {
+            Type = subProtocol == "graphql-ws" ? "start" : "subscribe",
+            Id = "123",
+            Payload = new GraphQLRequest
+            {
+                Query = "subscription { events { type message { id message from } } }",
+            },
+        });
+
+        await Task.Delay(1000);
```

Comment:
I would like to remove this delay, but the test then typically fails.  I believe this is because the prior call to `SendMessageAsync` gets processed more slowly than the following message via `AddMessageInternal`.  The latter call runs on a separate thread over a separate HttpClient, so there is no synchronization that ensures that the prior subscription request has completed before posting the new message.  Since the subscription server does not send any response, I cannot simply wait for a response here.  My only thought (which I have not tried) is that perhaps if some subsequent message that requires a response was sent via the WebSocket connection (e.g. CountMessages), then we could await that response before sending the add-message via a separate HttpClient.  Since there is no asynchronous code present in the chat service, this should eliminate the need for Task.Delay.  Still this is not a guaranteed fix because the BaseSubscriptionHandler is designed to process requests over the same WebSocket connection asynchronously. (See #821 )

Reply:
This same issue (necessary call to Task.Delay) is present elsewhere in this file and also in `TestServerExtensions.cs`

Reply:
Anyway it  would be better to place a comment

Reply:
ok i'll add a comment along the lines of what's written above

Reply:
```suggestion
        // Delay is necessary to allow time for the asynchronous websocket handler code
        // to execute prior to the independent call to AddMessageInternal below.
        // Since the websocket call does not return a response when the subscription
        // has completed being set up, there is no response we can await to determine
        // when to call AddMessageInternal.
        await Task.Delay(1000);
```

Reply:
is this sufficient?

Reply:
yes